# FUEL EFFICIENCY OF CARS

### Data wrangling and visualization
The fuel economy datasets are from 2008 and 2018. I will be conducting a wrangling process on the two datas seperately and then join/concatenate them.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

#loading datasets
df_08 = pd.read_csv('all_alpha_08.csv')#2008 fuel economy dataset
df_18 = pd.read_csv('all_alpha_18.csv')#2018 fuel economy dataset
df_08.head() #first 5 observations in 2008


Model  Displ      Cyl    Trans Drive      Fuel Sales Area Stnd  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   
1  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         FA   B5   
2  ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline         CA   U2   
3  ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline         FA   B5   
4   ACURA RL    3.5  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   

   Underhood ID    Veh Class Air Pollution Score FE Calc Appr City MPG  \
0  8HNXT03.7PKR          SUV                   7          Drv       15   
1  8HNXT03.7PKR          SUV                   6          Drv       15   
2  8HNXT02.3DKR          SUV                   7          Drv       17   
3  8HNXT02.3DKR          SUV                   6          Drv       17   
4  8HNXV03.5HKR  midsize car                   7          Drv       16   

  Hwy MPG Cmb MPG  Unadj Cmb MPG Greenhouse Gas Score SmartWay  
0      20      17        22.0527                    4       no  
1      20      17        22.0527                    4       no  
2      22      19        24.1745                    5       no  
3      22      19        24.1745                    5       no  
4      24      19        24.5629                    5       no

In [2]:
df_18.head() #first 5 observations in 2018

Model  Displ  Cyl       Trans Drive      Fuel Cert Region       Stnd  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA     T3B125   
1  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          CA         U2   
2  ACURA RDX    3.5  6.0  SemiAuto-6   4WD  Gasoline          FA     T3B125   
3  ACURA RDX    3.5  6.0  SemiAuto-6   4WD  Gasoline          CA         U2   
4  ACURA TLX    2.4  4.0       AMS-8   2WD  Gasoline          CA  L3ULEV125   

             Stnd Description  Underhood ID  Veh Class  Air Pollution Score  \
0      Federal Tier 3 Bin 125  JHNXT03.5GV3  small SUV                    3   
1      California LEV-II ULEV  JHNXT03.5GV3  small SUV                    3   
2      Federal Tier 3 Bin 125  JHNXT03.5GV3  small SUV                    3   
3      California LEV-II ULEV  JHNXT03.5GV3  small SUV                    3   
4  California LEV-III ULEV125  JHNXV02.4WH3  small car                    3   

  City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score SmartWay Comb CO2  
0       20      28      23                     5       No      386  
1       20      28      23                     5       No      386  
2       19      27      22                     4       No      402  
3       19      27      22                     4       No      402  
4       23      33      27                     6       No      330

In [3]:
#shape of both datasets
print( 'The 2008 dataset contains',df_08.shape[0], 'observations and', df_08.shape[1],'variables')
print( 'The 2018 dataset contains',df_18.shape[0], 'observations and', df_18.shape[1],'variables')

The 2008 dataset contains 2404 observations and 18 variables
The 2018 dataset contains 1611 observations and 18 variables


## DATA WRANGLING/ EXPLORATION

In [4]:
df_08.sample(10) #10 samples of the 2008 dataset

Model  Displ      Cyl    Trans Drive      Fuel Sales Area  \
1207    HYUNDAI Santa Fe    3.3  (6 cyl)  Auto-L5   2WD  Gasoline         FA   
360   CHEVROLET Colorado    2.9  (4 cyl)    Man-5   4WD  Gasoline         FA   
1724   MITSUBISHI Raider    3.7  (6 cyl)  Auto-L4   4WD  Gasoline         CA   
30               AUDI A3    2.0  (4 cyl)    Man-6   2WD  Gasoline         FA   
2362           VOLVO S60    2.5  (5 cyl)  Auto-S5   4WD  Gasoline         CA   
205            BMW 650ci    4.8  (8 cyl)  Auto-S6   2WD  Gasoline         FA   
423     CHEVROLET Malibu    2.4  (4 cyl)  Auto-L4   2WD  Gasoline         CA   
854            FORD F150    4.6  (8 cyl)  Auto-L4   4WD  Gasoline         FA   
1846          PONTIAC G6    3.6  (6 cyl)  Auto-S6   2WD  Gasoline         FA   
712          DODGE Nitro    4.0  (6 cyl)  Auto-L5   2WD  Gasoline         FA   

     Stnd  Underhood ID      Veh Class Air Pollution Score FE Calc Appr  \
1207   B5  8HYXT03.3PW5            SUV                   6          Drv   
360    B5  8GMXT02.9189         pickup                   6          Drv   
1724   U2  8CRXT0226P80         pickup                   7          Drv   
30     B5  8ADXV02.0366  station wagon                   6          Drv   
2362   U2  8VVXV02.5U2T      small car                   7          Drv   
205    B5  8BMXV04.8UL2      small car                   6          Drv   
423    L2  8GMXV02.4027    midsize car                   6          Drv   
854    B8  8FMXT04.63HB         pickup                   3          Drv   
1846   B5  8GMXB03.6148      small car                   6          Drv   
712    B5  8CRXT04.0NE0            SUV                   6          Drv   

     City MPG Hwy MPG Cmb MPG  Unadj Cmb MPG Greenhouse Gas Score SmartWay  
1207       17      24      20        25.5255                    6       no  
360        16      22      18        18.2998                    5       no  
1724       14      18      15        19.5386                    3       no  
30         21      29      24        31.1162                    7      yes  
2362       18      26      21        26.9019                    6      yes  
205        17      25      20        23.1800                    6       no  
423        22      30      25        24.6791                    7      yes  
854        13      17      14        18.6243                    2       no  
1846       17      26      20        20.0911                    6       no  
712        16      21      18        23.2000                    5       no

In [5]:
df_18.sample(10)  #10 samples of the 2018 dataset

Model  Displ  Cyl       Trans Drive      Fuel  \
1213   MINI Cooper Clubman All4    1.5  3.0  SemiAuto-8   4WD  Gasoline   
1061       MASERATI Ghibli S Q4    3.0  6.0      Auto-8   4WD  Gasoline   
864       JAGUAR F-Type S Coupe    3.0  6.0  SemiAuto-8   2WD  Gasoline   
1                     ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline   
1241  MINI Cooper S Convertible    2.0  4.0       Man-6   2WD  Gasoline   
670               GMC Sierra 15    5.3  8.0      Auto-6   4WD  Gasoline   
1233      MINI Cooper S Clubman    2.0  4.0       Man-6   2WD  Gasoline   
1158  MERCEDES-BENZ C300 4Matic    2.0  4.0      Auto-9   4WD  Gasoline   
1431       SUBARU Impreza Sport    2.0  4.0       Man-5   4WD  Gasoline   
210       BMW M240i Convertible    3.0  6.0  SemiAuto-8   4WD  Gasoline   

     Cert Region       Stnd            Stnd Description  Underhood ID  \
1213          FA     T3B125      Federal Tier 3 Bin 125  JBMXV01.5M3X   
1061          FA     T3B160      Federal Tier 3 Bin 160  JMAXV03.0DFI   
864           CA  L3SULEV30  California LEV-III SULEV30  JJLXJ03.0FSP   
1             CA         U2      California LEV-II ULEV  JHNXT03.5GV3   
1241          CA  L3SULEV30  California LEV-III SULEV30  JBMXV02.0M46   
670           CA  L3ULEV125  California LEV-III ULEV125  JGMXT05.3384   
1233          CA  L3SULEV30  California LEV-III SULEV30  JBMXV02.0M46   
1158          FA      T3B70       Federal Tier 3 Bin 70  JMBXJ02.0U2C   
1431          FA      T3B50       Federal Tier 3 Bin 50  JFJXJ02.0BUY   
210           CA  L3ULEV125  California LEV-III ULEV125  JBMXV03.0B2X   

            Veh Class  Air Pollution Score City MPG Hwy MPG Cmb MPG  \
1213      midsize car                    3       23      31      26   
1061      midsize car                    1       16      24      19   
864         small car                    7       19      27      22   
1           small SUV                    3       20      28      23   
1241        small car                    7       23      32      26   
670   special purpose                    3       15      19      17   
1233      midsize car                    7       22      31      25   
1158        small car                    5       22      30      25   
1431      midsize car                    6       23      31      26   
210         small car                    3       21      30      24   

      Greenhouse Gas Score SmartWay Comb CO2  
1213                     5       No      341  
1061                     3       No      475  
864                      4       No      407  
1                        5       No      386  
1241                     5       No      338  
670                      3       No      523  
1233                     5       No      349  
1158                     5       No      351  
1431                     5       No      337  
210                      5       No      365

In [6]:
df_08.info() #general info on the 2008 data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2404 entries, 0 to 2403
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model                 2404 non-null   object 
 1   Displ                 2404 non-null   float64
 2   Cyl                   2205 non-null   object 
 3   Trans                 2205 non-null   object 
 4   Drive                 2311 non-null   object 
 5   Fuel                  2404 non-null   object 
 6   Sales Area            2404 non-null   object 
 7   Stnd                  2404 non-null   object 
 8   Underhood ID          2404 non-null   object 
 9   Veh Class             2404 non-null   object 
 10  Air Pollution Score   2404 non-null   object 
 11  FE Calc Appr          2205 non-null   object 
 12  City MPG              2205 non-null   object 
 13  Hwy MPG               2205 non-null   object 
 14  Cmb MPG               2205 non-null   object 
 15  Unadj Cmb MPG        

In [7]:
df_18.info() #general info on the 2018 data 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1611 entries, 0 to 1610
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model                 1611 non-null   object 
 1   Displ                 1609 non-null   float64
 2   Cyl                   1609 non-null   float64
 3   Trans                 1611 non-null   object 
 4   Drive                 1611 non-null   object 
 5   Fuel                  1611 non-null   object 
 6   Cert Region           1611 non-null   object 
 7   Stnd                  1611 non-null   object 
 8   Stnd Description      1611 non-null   object 
 9   Underhood ID          1611 non-null   object 
 10  Veh Class             1611 non-null   object 
 11  Air Pollution Score   1611 non-null   int64  
 12  City MPG              1611 non-null   object 
 13  Hwy MPG               1611 non-null   object 
 14  Cmb MPG               1611 non-null   object 
 15  Greenhouse Gas Score 

In [8]:
df_08.isna().sum()

Model                     0
Displ                     0
Cyl                     199
Trans                   199
Drive                    93
Fuel                      0
Sales Area                0
Stnd                      0
Underhood ID              0
Veh Class                 0
Air Pollution Score       0
FE Calc Appr            199
City MPG                199
Hwy MPG                 199
Cmb MPG                 199
Unadj Cmb MPG           199
Greenhouse Gas Score    199
SmartWay                  0
dtype: int64

In [9]:
df_18.isna().sum()

Model                   0
Displ                   2
Cyl                     2
Trans                   0
Drive                   0
Fuel                    0
Cert Region             0
Stnd                    0
Stnd Description        0
Underhood ID            0
Veh Class               0
Air Pollution Score     0
City MPG                0
Hwy MPG                 0
Cmb MPG                 0
Greenhouse Gas Score    0
SmartWay                0
Comb CO2                0
dtype: int64

From the 2008 data set, we can see that there an errorneous inputation in the cyl column, it should be an int datatype but it looks like a str, some other features also have erroneous datatypes, which will be converted. There also seems to be a lot of null values, they are much and almost in more than 60% in a row, so we can't risk having manipulated values everywhere, I gotta drop them. Some variable don't look important, like the underhood ID, stnd, FE calc Appr, unadj cmb mpg (don't exactly know what that means), i'll have to drop those. There are 25 duplicate rows here.

Then for the 2018 dataset, the cyl variable seems to be a float, but should be an int(descrete),some other features also have erroneous datatypes, which will be converted. There are also null values, but just two(2), so I'll find a way to replace them with the appropriate values. Some variable also don't seem important and some are not present in the 2008 dataset, like stnd, stnd description, comb co2, underhood ID. 

Then in both datasets, the column names have spaces, so I'll rename then with underscores.

In [10]:
#checking for duplicated rows
sum(df_08.duplicated())

25

In [11]:
sum(df_18.duplicated())

0

There seems to be 25 duplicated rows in 2008 dataset

In [12]:
df_08.nunique()

Model                   436
Displ                    47
Cyl                       8
Trans                    14
Drive                     2
Fuel                      5
Sales Area                3
Stnd                     12
Underhood ID            343
Veh Class                 9
Air Pollution Score      13
FE Calc Appr              2
City MPG                 39
Hwy MPG                  43
Cmb MPG                  38
Unadj Cmb MPG           721
Greenhouse Gas Score     20
SmartWay                  2
dtype: int64

In [13]:
df_18.nunique()

Model                   367
Displ                    36
Cyl                       7
Trans                    26
Drive                     2
Fuel                      5
Cert Region               2
Stnd                     19
Stnd Description         19
Underhood ID            230
Veh Class                 9
Air Pollution Score       6
City MPG                 58
Hwy MPG                  62
Cmb MPG                  57
Greenhouse Gas Score     10
SmartWay                  3
Comb CO2                299
dtype: int64

Unique values are quite much for the `model` variable and analysing that could be a pain, so I'd like to group them better by creating a new column `make` housing the first word of all the `models`, that would be better. 

In [14]:
df_08['Fuel'].unique()

array(['Gasoline', 'ethanol/gas', 'ethanol', 'diesel', 'CNG'],
      dtype=object)

In [15]:
df_18['Fuel'].unique()

array(['Gasoline', 'Gasoline/Electricity', 'Diesel', 'Ethanol/Gas',
       'Electricity'], dtype=object)

The `fuel` column has hybrid fuel types like the `ethanol/gas and gasoline/electricity`. And also while merging these two datasets, I'm gonna have to use one letter case for them so they don't become confusing.

In [16]:
df_08['Cyl'].unique()

array(['(6 cyl)', '(4 cyl)', '(12 cyl)', '(8 cyl)', '(10 cyl)',
       '(16 cyl)', '(5 cyl)', nan, '(2 cyl)'], dtype=object)

In [17]:
df_18['Cyl'].unique()

array([ 6.,  4.,  5., 12., 16.,  8., nan,  3.])

In [18]:
df_08['Trans'].unique()

array(['Auto-S5', 'Man-6', 'Auto-S6', 'Auto-AV', 'Auto-S7', 'Auto-L4',
       'Auto-L6', 'Auto-4', 'Man-5', 'Auto-L5', nan, 'Auto-6', 'S8',
       'Auto-S4', 'Auto-L7'], dtype=object)

In [19]:
df_18['Trans'].unique()

array(['SemiAuto-6', 'AMS-8', 'SemiAuto-9', 'AutoMan-6', 'Auto-8',
       'AMS-6', 'AMS-7', 'Man-6', 'SemiAuto-8', 'SemiAuto-7', 'Auto-6',
       'Auto-10', 'SemiAuto-10', 'Man-7', 'Auto-9', 'CVT', 'Auto-4',
       'AutoMan-7', 'SCV-7', 'Auto-1', 'SCV-10', 'Auto-7', 'Man-5',
       'SCV-8', 'SCV-6', 'SemiAuto-5'], dtype=object)

Trans here has quite a large number of values. Analysing using this will be confusing, so I'd like to extract the first words in each of the word to make it easier.

In [20]:
df_08['Drive'].unique()

array(['4WD', '2WD', nan], dtype=object)

In [21]:
df_18['Drive'].unique()

array(['2WD', '4WD'], dtype=object)

In [22]:
df_08['Sales Area'].unique()

array(['CA', 'FA', 'FC'], dtype=object)

In [23]:
df_18['Cert Region'].unique()

array(['FA', 'CA'], dtype=object)

I might just have to narrow it down to a particular location.

In [24]:
df_08['Sales Area'].value_counts()

Sales Area
FA    1157
CA    1084
FC     163
Name: count, dtype: int64

In [25]:
df_18['Cert Region'].value_counts()

Cert Region
FA    813
CA    798
Name: count, dtype: int64

so I'll just go with data where the sales area/sales region is FA

In [26]:
df_08['Veh Class'].unique()

array(['SUV', 'midsize car', 'small car', 'station wagon', 'large car',
       'minivan', 'pickup', 'van', 'special purpose'], dtype=object)

In [27]:
df_18['Veh Class'].unique()

array(['small SUV', 'small car', 'midsize car', 'large car',
       'standard SUV', 'station wagon', 'pickup', 'special purpose',
       'minivan'], dtype=object)

In [28]:
df_08['Air Pollution Score'].unique()

array(['7', '6', '9.5', '3', '7/7', '6/6', '2', '3/3', '1', '0', '8', '9',
       '6/4'], dtype=object)

Due to the hybrid fuel types, some other variable also have dual observations, like the air pollution, the mpgs too. I wanted to explode the rows to two where these occur but that's gonna create a confusion which might look like they're different cars but infact are hybrids. I'd rather take the average of that as an int.

In [29]:
df_18['Air Pollution Score'].unique()

array([ 3,  1,  7,  5,  6, 10], dtype=int64)

In [30]:
df_08['SmartWay'].unique()

array(['no', 'yes'], dtype=object)

In [31]:
df_18['SmartWay'].unique()

array(['No', 'Yes', 'Elite'], dtype=object)

In [32]:
df_18['SmartWay'].value_counts()

SmartWay
No       1414
Yes       175
Elite      22
Name: count, dtype: int64

In [33]:
df_18[df_18['SmartWay'] =='Elite']

Model  Displ  Cyl       Trans Drive  \
160                            BMW 530e    2.0  4.0  SemiAuto-8   2WD   
161                            BMW 530e    2.0  4.0  SemiAuto-8   2WD   
162                            BMW 530e    2.0  4.0  SemiAuto-8   4WD   
163                            BMW 530e    2.0  4.0  SemiAuto-8   4WD   
456                      CHEVROLET Volt    1.5  4.0         CVT   2WD   
457                      CHEVROLET Volt    1.5  4.0         CVT   2WD   
605   FORD Fusion Energi Plug-in Hybrid    2.0  4.0         CVT   2WD   
606   FORD Fusion Energi Plug-in Hybrid    2.0  4.0         CVT   2WD   
996                   KIA Soul Electric    NaN  NaN      Auto-1   2WD   
997                   KIA Soul Electric    NaN  NaN      Auto-1   2WD   
1465             TOYOTA Camry Hybrid LE    2.5  4.0       SCV-6   2WD   
1466             TOYOTA Camry Hybrid LE    2.5  4.0       SCV-6   2WD   
1467         TOYOTA Camry Hybrid XLE/SE    2.5  4.0       SCV-6   2WD   
1468         TOYOTA Camry Hybrid XLE/SE    2.5  4.0       SCV-6   2WD   
1491                     TOYOTA Prius c    1.5  4.0         CVT   2WD   
1492                     TOYOTA Prius c    1.5  4.0         CVT   2WD   
1577                          VOLVO S90    2.0  4.0  SemiAuto-8   4WD   
1578                          VOLVO S90    2.0  4.0  SemiAuto-8   4WD   
1601                        VOLVO XC 60    2.0  4.0  SemiAuto-8   4WD   
1602                        VOLVO XC 60    2.0  4.0  SemiAuto-8   4WD   
1609                        VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD   
1610                        VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD   

                      Fuel Cert Region       Stnd            Stnd Description  \
160   Gasoline/Electricity          CA  L3SULEV30  California LEV-III SULEV30   
161   Gasoline/Electricity          FA      T3B30       Federal Tier 3 Bin 30   
162   Gasoline/Electricity          CA  L3SULEV30  California LEV-III SULEV30   
163   Gasoline/Electricity          FA      T3B30       Federal Tier 3 Bin 30   
456   Gasoline/Electricity          CA  L3SULEV30  California LEV-III SULEV30   
457   Gasoline/Electricity          FA      T3B30       Federal Tier 3 Bin 30   
605   Gasoline/Electricity          CA  L3SULEV30  California LEV-III SULEV30   
606   Gasoline/Electricity          FA      T3B30       Federal Tier 3 Bin 30   
996            Electricity          FA       T3B0        Federal Tier 3 Bin 0   
997            Electricity          CA        ZEV              California ZEV   
1465              Gasoline          CA  L3SULEV30  California LEV-III SULEV30   
1466              Gasoline          FA      T3B30       Federal Tier 3 Bin 30   
1467              Gasoline          CA  L3SULEV30  California LEV-III SULEV30   
1468              Gasoline          FA      T3B30       Federal Tier 3 Bin 30   
1491              Gasoline          CA  L3SULEV30  California LEV-III SULEV30   
1492              Gasoline          FA      T3B30       Federal Tier 3 Bin 30   
1577  Gasoline/Electricity          CA  L3SULEV30  California LEV-III SULEV30   
1578  Gasoline/Electricity          FA      T3B30       Federal Tier 3 Bin 30   
1601  Gasoline/Electricity          CA  L3SULEV30  California LEV-III SULEV30   
1602  Gasoline/Electricity          FA      T3B30       Federal Tier 3 Bin 30   
1609  Gasoline/Electricity          CA  L3SULEV30  California LEV-III SULEV30   
1610  Gasoline/Electricity          FA      T3B30       Federal Tier 3 Bin 30   

      Underhood ID      Veh Class  Air Pollution Score City MPG Hwy MPG  \
160   JBMXV02.0H30      small car                    7    27/70   31/75   
161   JBMXV02.0H30      small car                    7    27/70   31/75   
162   JBMXV02.0H30      small car                    7    27/66   31/68   
163   JBMXV02.0H30      small car                    7    27/66   31/68   
456   JGMXV01.5030      small car                    7   43/113   42/99   
457   JGMXV01.5030      small car       

In [34]:
df_18[df_18['City MPG'].str.contains('/')] #observations containing hybrid fuel types

Model  Displ  Cyl       Trans Drive                  Fuel  \
108      BMW 330e    2.0  4.0  SemiAuto-8   2WD  Gasoline/Electricity   
109      BMW 330e    2.0  4.0  SemiAuto-8   2WD  Gasoline/Electricity   
160      BMW 530e    2.0  4.0  SemiAuto-8   2WD  Gasoline/Electricity   
161      BMW 530e    2.0  4.0  SemiAuto-8   2WD  Gasoline/Electricity   
162      BMW 530e    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
...           ...    ...  ...         ...   ...                   ...   
1578    VOLVO S90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
1601  VOLVO XC 60    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
1602  VOLVO XC 60    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
1609  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
1610  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   

     Cert Region       Stnd            Stnd Description  Underhood ID  \
108           CA  L3ULEV125  California LEV-III ULEV125  JBMXV02.0H48   
109           FA     T3B125      Federal Tier 3 Bin 125  JBMXV02.0H48   
160           CA  L3SULEV30  California LEV-III SULEV30  JBMXV02.0H30   
161           FA      T3B30       Federal Tier 3 Bin 30  JBMXV02.0H30   
162           CA  L3SULEV30  California LEV-III SULEV30  JBMXV02.0H30   
...          ...        ...                         ...           ...   
1578          FA      T3B30       Federal Tier 3 Bin 30  JVVXV02.0P30   
1601          CA  L3SULEV30  California LEV-III SULEV30  JVVXJ02.0P30   
1602          FA      T3B30       Federal Tier 3 Bin 30  JVVXJ02.0P30   
1609          CA  L3SULEV30  California LEV-III SULEV30  JVVXJ02.0P30   
1610          FA      T3B30       Federal Tier 3 Bin 30  JVVXJ02.0P30   

         Veh Class  Air Pollution Score City MPG Hwy MPG Cmb MPG  \
108      small car                    3    28/66   34/78   30/71   
109      small car                    3    28/66   34/78   30/71   
160      small car                    7    27/70   31/75   29/72   
161      small car                    7    27/70   31/75   29/72   
162      small car                    7    27/66   31/68   28/67   
...            ...                  ...      ...     ...     ...   
1578   midsize car                    7    26/70   33/72   29/71   
1601     small SUV                    7    26/60   28/58   26/59   
1602     small SUV                    7    26/60   28/58   26/59   
1609  standard SUV                    7    26/63   30/61   27/62   
1610  standard SUV                    7    26/63   30/61   27/62   

      Greenhouse Gas Score SmartWay Comb CO2  
108                     10      Yes      189  
109                     10      Yes      189  
160                     10    Elite      193  
161                     10    Elite      193  
162                     10    Elite      200  
...                    ...      ...      ...  
1578                    10    Elite      161  
1601                    10    Elite      199  
1602                    10    Elite      199  
1609                    10    Elite      187  
1610                    10    Elite      187  

[79 rows x 18 columns]

In [35]:
df_08.describe()

Displ  Unadj Cmb MPG
count  2404.000000    2205.000000
mean      3.748918      23.916104
std       1.335785       6.366170
min       1.300000      10.018400
25%       2.500000      19.113900
50%       3.500000      23.921300
75%       4.800000      27.869300
max       8.400000      65.777800

In [36]:
df_18.describe()

Displ          Cyl  Air Pollution Score  Greenhouse Gas Score
count  1609.000000  1609.000000          1611.000000           1611.000000
mean      3.055687     5.479180             3.958411              4.711359
std       1.344574     1.749121             1.824303              1.657429
min       1.200000     3.000000             1.000000              1.000000
25%       2.000000     4.000000             3.000000              4.000000
50%       3.000000     6.000000             3.000000              5.000000
75%       3.600000     6.000000             5.000000              6.000000
max       8.000000    16.000000            10.000000             10.000000

some variables are not here because they're supposed to be quantitative but are recorded as qualitative.

### CLEANING PROCESS

In [37]:
df1 = df_08.copy() #a copy of 2008 dataset
df2 = df_18.copy() #a copy of 2018 dataset

In [38]:
#droping columns in 2008
L = ['Stnd', 'Underhood ID', 'FE Calc Appr', 'Unadj Cmb MPG','Sales Area']
df1.drop(L, axis = 1, inplace = True)
df1.columns


Index(['Model', 'Displ', 'Cyl', 'Trans', 'Drive', 'Fuel', 'Veh Class',
       'Air Pollution Score', 'City MPG', 'Hwy MPG', 'Cmb MPG',
       'Greenhouse Gas Score', 'SmartWay'],
      dtype='object')

In [39]:
#drop columns in 2018
M = ['Stnd', 'Stnd Description', 'Underhood ID', 'Comb CO2','Cert Region']
df2.drop( M, axis = 1, inplace = True)
df2.columns


Index(['Model', 'Displ', 'Cyl', 'Trans', 'Drive', 'Fuel', 'Veh Class',
       'Air Pollution Score', 'City MPG', 'Hwy MPG', 'Cmb MPG',
       'Greenhouse Gas Score', 'SmartWay'],
      dtype='object')

In [40]:
# replace spaces with underscores and lowercase labels for 2008 dataset
df1.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace=True)

# replace spaces with underscores and lowercase labels for 2008 dataset
df2.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace=True)
# confirm changes
(df1.columns == df2.columns).all() #checking if the columns match for both datasets

True

In [41]:
#check for duplicated rows in 2008
df1.duplicated().sum()

653

In [42]:
#checking for duplicated rows in 2018
df2.duplicated().sum()

778

In [43]:
#dropping duplicated rows in both years
df1.drop_duplicates(inplace = True)
df2.drop_duplicates(inplace = True)

print(df1.duplicated().sum()) #checking for duplicates
print(df2.duplicated().sum())

0
0


In [44]:
df1.isnull().sum() #checking again for null values

model                     0
displ                     0
cyl                     101
trans                   101
drive                    40
fuel                      0
veh_class                 0
air_pollution_score       0
city_mpg                101
hwy_mpg                 101
cmb_mpg                 101
greenhouse_gas_score    101
smartway                  0
dtype: int64

101 missing values doesn't look good. I can't replace them because it will be too much of guess data, so i have to drop them. Drive column has 40 missing values and its part of the rows that have missing values.

In [45]:
#looking at the drive column
df1[df1['cyl'].isnull()]

model  displ  cyl trans drive      fuel veh_class  \
389   CHEVROLET Express 3500    6.0  NaN   NaN   NaN  Gasoline       van   
390   CHEVROLET Express 3500    6.2  NaN   NaN   NaN  Gasoline       van   
391   CHEVROLET Express 3500    4.8  NaN   NaN   NaN  Gasoline       van   
394   CHEVROLET Express 3500    6.0  NaN   NaN   NaN  Gasoline       van   
449   CHEVROLET Silverado 25    4.8  NaN   NaN   2WD  Gasoline    pickup   
...                      ...    ...  ...   ...   ...       ...       ...   
1463       LINCOLN Navigator    5.4  NaN   NaN   4WD  Gasoline       SUV   
1465       LINCOLN Navigator    5.4  NaN   NaN   4WD  Gasoline       SUV   
1934     PORSCHE Cayenne GTS    4.8  NaN   NaN   4WD  Gasoline       SUV   
1935     PORSCHE Cayenne GTS    4.8  NaN   NaN   4WD  Gasoline       SUV   
1939   PORSCHE Cayenne Turbo    4.8  NaN   NaN   4WD  Gasoline       SUV   

     air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score  \
389                    3      NaN     NaN     NaN                  NaN   
390                    3      NaN     NaN     NaN                  NaN   
391                    2      NaN     NaN     NaN                  NaN   
394                    2      NaN     NaN     NaN                  NaN   
449                    2      NaN     NaN     NaN                  NaN   
...                  ...      ...     ...     ...                  ...   
1463                   7      NaN     NaN     NaN                  NaN   
1465                   6      NaN     NaN     NaN                  NaN   
1934                   7      NaN     NaN     NaN                  NaN   
1935                   6      NaN     NaN     NaN                  NaN   
1939                   6      NaN     NaN     NaN                  NaN   

     smartway  
389        no  
390        no  
391        no  
394        no  
449        no  
...       ...  
1463       no  
1465       no  
1934       no  
1935       no  
1939       no  

[101 rows x 13 columns]

In [46]:
#dropping missing values
df1.dropna(axis = 0, inplace = True)
df1.isnull().sum().any() #checking for missing values

False

In [47]:
df2.isna().sum() #checking for missing values in 2018

model                   0
displ                   1
cyl                     1
trans                   0
drive                   0
fuel                    0
veh_class               0
air_pollution_score     0
city_mpg                0
hwy_mpg                 0
cmb_mpg                 0
greenhouse_gas_score    0
smartway                0
dtype: int64

there are 2, one in displ and the other in cyl, so i'll fill them up with the mean values

In [48]:
df2[df2['displ'].isnull()]

model  displ  cyl   trans drive         fuel      veh_class  \
996  KIA Soul Electric    NaN  NaN  Auto-1   2WD  Electricity  station wagon   

     air_pollution_score city_mpg hwy_mpg cmb_mpg  greenhouse_gas_score  \
996                   10      124      93     108                    10   

    smartway  
996    Elite

In [49]:
#replacing the null values with the mean of the cyl and displ
df2['displ'].fillna(df2['displ'].mean(), inplace = True)
df2['cyl'].fillna(df2['cyl'].mean(), inplace = True)

df2.isna().sum().any()

False

##### converting the cyl column in 2008 from str to int

In [50]:
df1['cyl'].value_counts()

cyl
(6 cyl)     646
(4 cyl)     455
(8 cyl)     392
(5 cyl)     109
(12 cyl)     30
(10 cyl)     15
(2 cyl)       2
(16 cyl)      1
Name: count, dtype: int64

In [51]:
df1['cyl'] = df1['cyl'].str.extract('(\d+)').astype(int)
df1['cyl'].value_counts()

cyl
6     646
4     455
8     392
5     109
12     30
10     15
2       2
16      1
Name: count, dtype: int64

In [52]:
#converting cyl in 2018 from float to int dtype in cyl 2018
df2['cyl'] = df2['cyl'].astype(int)
df2['cyl'].dtype

dtype('int32')

### Fixing hybrid values in fuel, air_pollution_score, the mpgs and greenhouse_gas_score

In [53]:
#first we have to check the fuel types that contain 2 values

df1[df1['fuel'].str.contains('/')] #2008 dataset

model  displ  cyl    trans drive         fuel  \
316   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   2WD  ethanol/gas   
317   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   4WD  ethanol/gas   
321   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   2WD  ethanol/gas   
322   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   4WD  ethanol/gas   
380     CHEVROLET Express 1500    5.3    8  Auto-L4   4WD  ethanol/gas   
...                        ...    ...  ...      ...   ...          ...   
1734         MITSUBISHI Raider    4.7    8  Auto-L5   4WD  ethanol/gas   
1764             NISSAN Armada    5.6    8  Auto-L5   2WD  ethanol/gas   
1765             NISSAN Armada    5.6    8  Auto-L5   4WD  ethanol/gas   
1808              NISSAN Titan    5.6    8  Auto-L5   2WD  ethanol/gas   
1809              NISSAN Titan    5.6    8  Auto-L5   4WD  ethanol/gas   

     veh_class air_pollution_score city_mpg hwy_mpg cmb_mpg  \
316        SUV                 7/7    11/14   15/20   12/16   
317        SUV                 7/7    11/14   14/19   12/16   
321        SUV                 6/6    11/14   15/20   12/16   
322        SUV                 6/6    11/14   14/19   12/16   
380        van                 6/6     9/12   12/16   10/14   
...        ...                 ...      ...     ...     ...   
1734    pickup                 3/3     9/14   12/19   10/15   
1764       SUV                 6/6     9/12   13/18   10/14   
1765       SUV                 6/6     9/12   13/17   10/14   
1808    pickup                 6/6     9/12   13/17   10/14   
1809    pickup                 6/6     9/12   12/17   10/14   

     greenhouse_gas_score smartway  
316                   6/4       no  
317                   6/4       no  
321                   6/4       no  
322                   6/4       no  
380                   4/2       no  
...                   ...      ...  
1734                  4/3       no  
1764                  4/2       no  
1765                  4/2       no  
1808                  4/2       no  
1809                  4/2       no  

[65 rows x 13 columns]

In [54]:
df2[df2['fuel'].str.contains('/')]

model  displ  cyl       trans drive  \
108                            BMW 330e    2.0    4  SemiAuto-8   2WD   
160                            BMW 530e    2.0    4  SemiAuto-8   2WD   
162                            BMW 530e    2.0    4  SemiAuto-8   4WD   
188                            BMW 740e    2.0    4  SemiAuto-8   4WD   
382                    CHEVROLET Impala    3.6    6  SemiAuto-6   2WD   
394              CHEVROLET Silverado 15    4.3    6      Auto-6   2WD   
396              CHEVROLET Silverado 15    4.3    6      Auto-6   4WD   
398              CHEVROLET Silverado 15    5.3    8      Auto-6   2WD   
428             CHEVROLET Suburban 1500    5.3    8      Auto-6   2WD   
432             CHEVROLET Suburban 1500    5.3    8      Auto-6   4WD   
436                CHEVROLET Tahoe 1500    5.3    8      Auto-6   2WD   
440                CHEVROLET Tahoe 1500    5.3    8      Auto-6   4WD   
454                      CHEVROLET Volt    1.5    4         CVT   2WD   
456                      CHEVROLET Volt    1.5    4         CVT   2WD   
458                        CHRYSLER 300    3.6    6      Auto-8   2WD   
462                        CHRYSLER 300    3.6    6      Auto-8   4WD   
492                       DODGE Charger    3.6    6      Auto-8   2WD   
496                       DODGE Charger    3.6    6      Auto-8   4WD   
520                       DODGE Journey    3.6    6      Auto-6   2WD   
605   FORD Fusion Energi Plug-in Hybrid    2.0    4         CVT   2WD   
609                FORD Transit Connect    2.5    4  SemiAuto-6   2WD   
610                FORD Transit Connect    2.5    4  SemiAuto-6   2WD   
659                       GMC Sierra 15    4.3    6      Auto-6   2WD   
661                       GMC Sierra 15    4.3    6      Auto-6   4WD   
663                       GMC Sierra 15    5.3    8      Auto-6   2WD   
697                      GMC Yukon 1500    5.3    8      Auto-6   2WD   
701                      GMC Yukon 1500    5.3    8      Auto-6   4WD   
709                   GMC Yukon 1500 XL    5.3    8      Auto-6   2WD   
715                   GMC Yukon XL 1500    5.3    8      Auto-6   4WD   
892                       JEEP Cherokee    2.4    4      Auto-9   2WD   
896                       JEEP Cherokee    2.4    4      Auto-9   4WD   
932                        KARMA Revero    2.0    4      Auto-1   2WD   
1162        MERCEDES-BENZ CLA250 4Matic    2.0    4   AutoMan-7   4WD   
1179        MERCEDES-BENZ GLA250 4Matic    2.0    4   AutoMan-7   4WD   
1191        MERCEDES-BENZ GLE350 4Matic    3.5    6      Auto-7   4WD   
1192        MERCEDES-BENZ GLE350 4Matic    3.5    6      Auto-7   4WD   
1255     MINI Cooper SE Countryman All4    1.5    3  SemiAuto-6   4WD   
1507                 TOYOTA Sequoia FFV    5.7    8  SemiAuto-6   4WD   
1517                  TOYOTA Tundra FFV    5.7    8  SemiAuto-6   4WD   
1577                          VOLVO S90    2.0    4  SemiAuto-8   4WD   
1601                        VOLVO XC 60    2.0    4  SemiAuto-8   4WD   
1609                        VOLVO XC 90    2.0    4  SemiAuto-8   4WD   

                      fuel        veh_class  air_pollution_score city_mpg  \
108   Gasoline/Electricity        small car                    3    28/66   
160   Gasoline/Electricity        small car                    7    27/70   
162   Gasoline/Electricity        small car                    7    27/66   
188   Gasoline/Electricity        large car                    3    25/62   
382            Ethanol/Gas        large car                    5    14/18   
394            Ethanol/Gas           pickup                    5    12/18   
396            Ethanol/Gas           pickup                    5    12/17   
398            Ethanol/Gas           pickup                    3    12/16   
428            Ethanol/Gas     standard SUV                    3    12/16   
432            Ethanol/Gas     standard SUV                    3    11/16   
436            Ethanol/Gas     standard SUV    

For the cleaning of this part i've decided to leave the `fuel` column, since it's a string and will remain as such, but for other columns which are meant to be numerical and are affected by this hybrid fuel types, I'll take their average so as to keep the columns analyzable. so first I'll split the values, add them up, fine the mean and convert to int. Tadaaa

In [55]:
hbd = ['air_pollution_score','city_mpg','hwy_mpg','cmb_mpg','greenhouse_gas_score'] #hybrid columns
for i in hbd:
    print((df1[i].dtype))

object
object
object
object
object


In [56]:
df1['air_pollution_score'].unique()

array(['7', '6', '9.5', '3', '7/7', '6/6', '3/3', '8', '9', '1', '6/4'],
      dtype=object)

In [57]:
#splitting the 2008 hybrid columns
#split the values and calculate the average
for i in hbd:
    df1[i] = df1[i].str.split('/').apply(lambda x: int((int(x[0]) + int(x[1]))/2) if len(x) == 2 else (x[0]))

df1[df1['fuel'].str.contains('/')]

model  displ  cyl    trans drive         fuel  \
316   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   2WD  ethanol/gas   
317   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   4WD  ethanol/gas   
321   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   2WD  ethanol/gas   
322   CHEVROLET Avalanche 1500    5.3    8  Auto-L4   4WD  ethanol/gas   
380     CHEVROLET Express 1500    5.3    8  Auto-L4   4WD  ethanol/gas   
...                        ...    ...  ...      ...   ...          ...   
1734         MITSUBISHI Raider    4.7    8  Auto-L5   4WD  ethanol/gas   
1764             NISSAN Armada    5.6    8  Auto-L5   2WD  ethanol/gas   
1765             NISSAN Armada    5.6    8  Auto-L5   4WD  ethanol/gas   
1808              NISSAN Titan    5.6    8  Auto-L5   2WD  ethanol/gas   
1809              NISSAN Titan    5.6    8  Auto-L5   4WD  ethanol/gas   

     veh_class air_pollution_score city_mpg hwy_mpg cmb_mpg  \
316        SUV                   7       12      17      14   
317        SUV                   7       12      16      14   
321        SUV                   6       12      17      14   
322        SUV                   6       12      16      14   
380        van                   6       10      14      12   
...        ...                 ...      ...     ...     ...   
1734    pickup                   3       11      15      12   
1764       SUV                   6       10      15      12   
1765       SUV                   6       10      15      12   
1808    pickup                   6       10      15      12   
1809    pickup                   6       10      14      12   

     greenhouse_gas_score smartway  
316                     5       no  
317                     5       no  
321                     5       no  
322                     5       no  
380                     3       no  
...                   ...      ...  
1734                    3       no  
1764                    3       no  
1765                    3       no  
1808                    3       no  
1809                    3       no  

[65 rows x 13 columns]

In [58]:
#splitting the 2018 hybrid columns
hbd2 = ['city_mpg','hwy_mpg','cmb_mpg'] #hybrid columns

#split the values and calculate the average
for i in hbd2:
    df2[i] = df2[i].str.split('/').apply(lambda x: int((int(x[0]) + int(x[1]))/2) if len(x) == 2 else (x[0]))

df2[df2['fuel'].str.contains('/')]

model  displ  cyl       trans drive  \
108                            BMW 330e    2.0    4  SemiAuto-8   2WD   
160                            BMW 530e    2.0    4  SemiAuto-8   2WD   
162                            BMW 530e    2.0    4  SemiAuto-8   4WD   
188                            BMW 740e    2.0    4  SemiAuto-8   4WD   
382                    CHEVROLET Impala    3.6    6  SemiAuto-6   2WD   
394              CHEVROLET Silverado 15    4.3    6      Auto-6   2WD   
396              CHEVROLET Silverado 15    4.3    6      Auto-6   4WD   
398              CHEVROLET Silverado 15    5.3    8      Auto-6   2WD   
428             CHEVROLET Suburban 1500    5.3    8      Auto-6   2WD   
432             CHEVROLET Suburban 1500    5.3    8      Auto-6   4WD   
436                CHEVROLET Tahoe 1500    5.3    8      Auto-6   2WD   
440                CHEVROLET Tahoe 1500    5.3    8      Auto-6   4WD   
454                      CHEVROLET Volt    1.5    4         CVT   2WD   
456                      CHEVROLET Volt    1.5    4         CVT   2WD   
458                        CHRYSLER 300    3.6    6      Auto-8   2WD   
462                        CHRYSLER 300    3.6    6      Auto-8   4WD   
492                       DODGE Charger    3.6    6      Auto-8   2WD   
496                       DODGE Charger    3.6    6      Auto-8   4WD   
520                       DODGE Journey    3.6    6      Auto-6   2WD   
605   FORD Fusion Energi Plug-in Hybrid    2.0    4         CVT   2WD   
609                FORD Transit Connect    2.5    4  SemiAuto-6   2WD   
610                FORD Transit Connect    2.5    4  SemiAuto-6   2WD   
659                       GMC Sierra 15    4.3    6      Auto-6   2WD   
661                       GMC Sierra 15    4.3    6      Auto-6   4WD   
663                       GMC Sierra 15    5.3    8      Auto-6   2WD   
697                      GMC Yukon 1500    5.3    8      Auto-6   2WD   
701                      GMC Yukon 1500    5.3    8      Auto-6   4WD   
709                   GMC Yukon 1500 XL    5.3    8      Auto-6   2WD   
715                   GMC Yukon XL 1500    5.3    8      Auto-6   4WD   
892                       JEEP Cherokee    2.4    4      Auto-9   2WD   
896                       JEEP Cherokee    2.4    4      Auto-9   4WD   
932                        KARMA Revero    2.0    4      Auto-1   2WD   
1162        MERCEDES-BENZ CLA250 4Matic    2.0    4   AutoMan-7   4WD   
1179        MERCEDES-BENZ GLA250 4Matic    2.0    4   AutoMan-7   4WD   
1191        MERCEDES-BENZ GLE350 4Matic    3.5    6      Auto-7   4WD   
1192        MERCEDES-BENZ GLE350 4Matic    3.5    6      Auto-7   4WD   
1255     MINI Cooper SE Countryman All4    1.5    3  SemiAuto-6   4WD   
1507                 TOYOTA Sequoia FFV    5.7    8  SemiAuto-6   4WD   
1517                  TOYOTA Tundra FFV    5.7    8  SemiAuto-6   4WD   
1577                          VOLVO S90    2.0    4  SemiAuto-8   4WD   
1601                        VOLVO XC 60    2.0    4  SemiAuto-8   4WD   
1609                        VOLVO XC 90    2.0    4  SemiAuto-8   4WD   

                      fuel        veh_class  air_pollution_score city_mpg  \
108   Gasoline/Electricity        small car                    3       47   
160   Gasoline/Electricity        small car                    7       48   
162   Gasoline/Electricity        small car                    7       46   
188   Gasoline/Electricity        large car                    3       43   
382            Ethanol/Gas        large car                    5       16   
394            Ethanol/Gas           pickup                    5       15   
396            Ethanol/Gas           pickup                    5       14   
398            Ethanol/Gas           pickup                    3       14   
428            Ethanol/Gas     standard SUV                    3       14   
432            Ethanol/Gas     standard SUV                    3       13   
436            Ethanol/Gas     standard SUV    

#### changing the datatypes of some columns

In [62]:
print(df1.info())
print(df2.info())

<class 'pandas.core.frame.DataFrame'>
Index: 1650 entries, 0 to 2403
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   model                 1650 non-null   object 
 1   displ                 1650 non-null   float64
 2   cyl                   1650 non-null   int32  
 3   trans                 1650 non-null   object 
 4   drive                 1650 non-null   object 
 5   fuel                  1650 non-null   object 
 6   veh_class             1650 non-null   object 
 7   air_pollution_score   1650 non-null   object 
 8   city_mpg              1650 non-null   object 
 9   hwy_mpg               1650 non-null   object 
 10  cmb_mpg               1650 non-null   object 
 11  greenhouse_gas_score  1650 non-null   object 
 12  smartway              1650 non-null   object 
dtypes: float64(1), int32(1), object(11)
memory usage: 174.0+ KB
None
<class 'pandas.core.frame.DataFrame'>
Index: 833 entries, 0 to 

In [63]:
#replace a numeric value with the int part

df1['air_pollution_score'].replace('9.5', '9', inplace = True)
X = ['air_pollution_score','city_mpg','hwy_mpg','cmb_mpg','greenhouse_gas_score']
for i in X:
    df1[i] = df1[i].astype(int)

y = ['city_mpg','hwy_mpg','cmb_mpg']
for j in y:
    df2[j] = df2[j].astype(int)

In [64]:
#checking results
print(df1.info())
print(df2.info())

<class 'pandas.core.frame.DataFrame'>
Index: 1650 entries, 0 to 2403
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   model                 1650 non-null   object 
 1   displ                 1650 non-null   float64
 2   cyl                   1650 non-null   int32  
 3   trans                 1650 non-null   object 
 4   drive                 1650 non-null   object 
 5   fuel                  1650 non-null   object 
 6   veh_class             1650 non-null   object 
 7   air_pollution_score   1650 non-null   int32  
 8   city_mpg              1650 non-null   int32  
 9   hwy_mpg               1650 non-null   int32  
 10  cmb_mpg               1650 non-null   int32  
 11  greenhouse_gas_score  1650 non-null   int32  
 12  smartway              1650 non-null   object 
dtypes: float64(1), int32(6), object(6)
memory usage: 141.8+ KB
None
<class 'pandas.core.frame.DataFrame'>
Index: 833 entries, 0 to 1

In [65]:
df1['year'] = 2008
df2['year'] = 2018

df1.head(1)
df2.head(1)

model  displ  cyl       trans drive      fuel  veh_class  \
0  ACURA RDX    3.5    6  SemiAuto-6   2WD  Gasoline  small SUV   

   air_pollution_score  city_mpg  hwy_mpg  cmb_mpg  greenhouse_gas_score  \
0                    3        20       28       23                     5   

  smartway  year  
0       No  2018

##### Creating a make column from the model column for better grouping during analysis

In [66]:
df1['make'] = df1['model'].str.split().str[0]
df2['make'] = df2['model'].str.split().str[0]

df1.head()

model  displ  cyl    trans drive      fuel    veh_class  \
0  ACURA MDX    3.7    6  Auto-S5   4WD  Gasoline          SUV   
1  ACURA MDX    3.7    6  Auto-S5   4WD  Gasoline          SUV   
2  ACURA RDX    2.3    4  Auto-S5   4WD  Gasoline          SUV   
3  ACURA RDX    2.3    4  Auto-S5   4WD  Gasoline          SUV   
4   ACURA RL    3.5    6  Auto-S5   4WD  Gasoline  midsize car   

   air_pollution_score  city_mpg  hwy_mpg  cmb_mpg  greenhouse_gas_score  \
0                    7        15       20       17                     4   
1                    6        15       20       17                     4   
2                    7        17       22       19                     5   
3                    6        17       22       19                     5   
4                    7        16       24       19                     5   

  smartway  year   make  
0       no  2008  ACURA  
1       no  2008  ACURA  
2       no  2008  ACURA  
3       no  2008  ACURA  
4       no  2008  ACURA

In [67]:
df2.head()

model  displ  cyl       trans drive      fuel  veh_class  \
0  ACURA RDX    3.5    6  SemiAuto-6   2WD  Gasoline  small SUV   
2  ACURA RDX    3.5    6  SemiAuto-6   4WD  Gasoline  small SUV   
4  ACURA TLX    2.4    4       AMS-8   2WD  Gasoline  small car   
6  ACURA TLX    3.5    6  SemiAuto-9   2WD  Gasoline  small car   
8  ACURA TLX    3.5    6  SemiAuto-9   4WD  Gasoline  small car   

   air_pollution_score  city_mpg  hwy_mpg  cmb_mpg  greenhouse_gas_score  \
0                    3        20       28       23                     5   
2                    3        19       27       22                     4   
4                    3        23       33       27                     6   
6                    3        20       32       24                     5   
8                    3        21       30       24                     5   

  smartway  year   make  
0       No  2018  ACURA  
2       No  2018  ACURA  
4       No  2018  ACURA  
6       No  2018  ACURA  
8       No  2018  ACURA

- missing values in 2008, and in 2018@
- 25 duplicated observations in 2008 @
- space in column titles@
- irrelevant columns@
- inconsistent values of Cyl(2008)@
- erroneous datatypes for cyl(int)2008/18, 2008vAir pollution score(int), city_, hwy_, cmb_mpgs, GHG(int)@
- create new columns for year and make@
- 2008 smartway column change values to "Yes,No" not yes,no
- 2018 smartway replace elite to yes@
- columns that contain hybrid values separated with `slash(/)@`

#### Merge dataframes

In [68]:
merged_df = pd.concat([df1, df2], ignore_index = True)
merged_df.head()

model  displ  cyl    trans drive      fuel    veh_class  \
0  ACURA MDX    3.7    6  Auto-S5   4WD  Gasoline          SUV   
1  ACURA MDX    3.7    6  Auto-S5   4WD  Gasoline          SUV   
2  ACURA RDX    2.3    4  Auto-S5   4WD  Gasoline          SUV   
3  ACURA RDX    2.3    4  Auto-S5   4WD  Gasoline          SUV   
4   ACURA RL    3.5    6  Auto-S5   4WD  Gasoline  midsize car   

   air_pollution_score  city_mpg  hwy_mpg  cmb_mpg  greenhouse_gas_score  \
0                    7        15       20       17                     4   
1                    6        15       20       17                     4   
2                    7        17       22       19                     5   
3                    6        17       22       19                     5   
4                    7        16       24       19                     5   

  smartway  year   make  
0       no  2008  ACURA  
1       no  2008  ACURA  
2       no  2008  ACURA  
3       no  2008  ACURA  
4       no  2008  ACURA

In [69]:
merged_df.shape

(2483, 15)

In [70]:
#lets look at the trans unique values
merged_df['trans'].unique()

array(['Auto-S5', 'Man-6', 'Auto-S6', 'Auto-AV', 'Auto-S7', 'Auto-L4',
       'Auto-L6', 'Auto-4', 'Man-5', 'Auto-L5', 'Auto-6', 'S8', 'Auto-S4',
       'Auto-L7', 'SemiAuto-6', 'AMS-8', 'SemiAuto-9', 'AutoMan-6',
       'Auto-8', 'AMS-6', 'AMS-7', 'SemiAuto-8', 'SemiAuto-7', 'Auto-10',
       'SemiAuto-10', 'Man-7', 'Auto-9', 'CVT', 'AutoMan-7', 'SCV-7',
       'Auto-1', 'SCV-10', 'Auto-7', 'SCV-8', 'SCV-6', 'SemiAuto-5'],
      dtype=object)

That's a lot. I'd only want trans_types though
##### Extracting the first words in the transmission column as transmission type

In [71]:
merged_df['trans_type'] = merged_df['trans'].str.split('-').str[0]

In [72]:
merged_df['trans_type'].unique()

array(['Auto', 'Man', 'S8', 'SemiAuto', 'AMS', 'AutoMan', 'CVT', 'SCV'],
      dtype=object)

Auto: Automatic transmission, Man: Manual transmission, SemiAuto: Semi-automatic transmission, CVT: Continuously Variable Transmission, AMS: Automated Manual Transmission, SCV: Single-Clutch Variable Transmission, AutoMan: Automatic and manual transmission. `S8` doesn't really fit in right? I'll have to replace that with Auto, because looking at the 2008 dataset, it seems that's where it belonged to.

In [73]:
merged_df[merged_df['trans_type'] =='S8']

model  displ  cyl trans drive      fuel    veh_class  \
913    LEXUS GS 460    4.6    8    S8   2WD  Gasoline  midsize car   
914    LEXUS GS 460    4.6    8    S8   2WD  Gasoline  midsize car   
925    LEXUS LS 460    4.6    8    S8   2WD  Gasoline  midsize car   
926    LEXUS LS 460    4.6    8    S8   2WD  Gasoline  midsize car   
927  LEXUS LS 460 L    4.6    8    S8   2WD  Gasoline  midsize car   
928  LEXUS LS 460 L    4.6    8    S8   2WD  Gasoline  midsize car   
929  LEXUS LS 600HL    5.0    8    S8   4WD  Gasoline  midsize car   
930  LEXUS LS 600HL    5.0    8    S8   4WD  Gasoline  midsize car   

     air_pollution_score  city_mpg  hwy_mpg  cmb_mpg  greenhouse_gas_score  \
913                    7        17       24       20                     6   
914                    6        17       24       20                     6   
925                    7        16       24       19                     5   
926                    6        16       24       19                     5   
927                    7        16       24       19                     5   
928                    6        16       24       19                     5   
929                    9        20       22       21                     6   
930                    8        20       22       21                     6   

    smartway  year   make trans_type  
913      yes  2008  LEXUS         S8  
914       no  2008  LEXUS         S8  
925       no  2008  LEXUS         S8  
926       no  2008  LEXUS         S8  
927       no  2008  LEXUS         S8  
928       no  2008  LEXUS         S8  
929      yes  2008  LEXUS         S8  
930      yes  2008  LEXUS         S8

In [74]:
merged_df['trans_type'].replace('S8','Auto', inplace = True)
merged_df[merged_df['trans_type'] =='S8']

Empty DataFrame
Columns: [model, displ, cyl, trans, drive, fuel, veh_class, air_pollution_score, city_mpg, hwy_mpg, cmb_mpg, greenhouse_gas_score, smartway, year, make, trans_type]
Index: []

In [75]:
df = ['model', 'make', 'year', 'displ','cyl', 'trans','trans_type','drive','fuel',
                      'veh_class', 'city_mpg', 'hwy_mpg', 'cmb_mpg', 'air_pollution_score',
                       'greenhouse_gas_score','smartway']
merged_df = merged_df.reindex(columns = df)
merged_df.head(10)

model   make  year  displ  cyl    trans trans_type drive      fuel  \
0  ACURA MDX  ACURA  2008    3.7    6  Auto-S5       Auto   4WD  Gasoline   
1  ACURA MDX  ACURA  2008    3.7    6  Auto-S5       Auto   4WD  Gasoline   
2  ACURA RDX  ACURA  2008    2.3    4  Auto-S5       Auto   4WD  Gasoline   
3  ACURA RDX  ACURA  2008    2.3    4  Auto-S5       Auto   4WD  Gasoline   
4   ACURA RL  ACURA  2008    3.5    6  Auto-S5       Auto   4WD  Gasoline   
5   ACURA RL  ACURA  2008    3.5    6  Auto-S5       Auto   4WD  Gasoline   
6   ACURA TL  ACURA  2008    3.2    6  Auto-S5       Auto   2WD  Gasoline   
7   ACURA TL  ACURA  2008    3.5    6  Auto-S5       Auto   2WD  Gasoline   
8   ACURA TL  ACURA  2008    3.5    6    Man-6        Man   2WD  Gasoline   
9   ACURA TL  ACURA  2008    3.2    6  Auto-S5       Auto   2WD  Gasoline   

     veh_class  city_mpg  hwy_mpg  cmb_mpg  air_pollution_score  \
0          SUV        15       20       17                    7   
1          SUV        15       20       17                    6   
2          SUV        17       22       19                    7   
3          SUV        17       22       19                    6   
4  midsize car        16       24       19                    7   
5  midsize car        16       24       19                    6   
6  midsize car        18       26       21                    7   
7  midsize car        17       26       20                    7   
8  midsize car        18       27       21                    7   
9  midsize car        18       26       21                    6   

   greenhouse_gas_score smartway  
0                     4       no  
1                     4       no  
2                     5       no  
3                     5       no  
4                     5       no  
5                     5       no  
6                     6      yes  
7                     6      yes  
8                     6      yes  
9                     6       no

In [76]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2483 entries, 0 to 2482
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   model                 2483 non-null   object 
 1   make                  2483 non-null   object 
 2   year                  2483 non-null   int64  
 3   displ                 2483 non-null   float64
 4   cyl                   2483 non-null   int32  
 5   trans                 2483 non-null   object 
 6   trans_type            2483 non-null   object 
 7   drive                 2483 non-null   object 
 8   fuel                  2483 non-null   object 
 9   veh_class             2483 non-null   object 
 10  city_mpg              2483 non-null   int32  
 11  hwy_mpg               2483 non-null   int32  
 12  cmb_mpg               2483 non-null   int32  
 13  air_pollution_score   2483 non-null   int64  
 14  greenhouse_gas_score  2483 non-null   int64  
 15  smartway             

In [77]:
merged_df.duplicated().sum()

16

In [78]:
#dropping duplicated
merged_df.drop_duplicates(inplace = True)
merged_df.duplicated().sum()

0

##### let us look at the descriptive analysis of the numerical values

In [79]:
merged_df.describe()

year        displ          cyl     city_mpg      hwy_mpg  \
count  2467.000000  2467.000000  2467.000000  2467.000000  2467.000000   
mean   2011.376571     3.373064     5.830563    18.621403    25.460478   
std       4.730068     1.292659     1.757059     5.768738     6.064966   
min    2008.000000     1.200000     2.000000     8.000000    13.000000   
25%    2008.000000     2.400000     4.000000    15.000000    22.000000   
50%    2008.000000     3.000000     6.000000    18.000000    25.000000   
75%    2018.000000     4.000000     6.000000    21.000000    29.000000   
max    2018.000000     8.400000    16.000000   124.000000    93.000000   

           cmb_mpg  air_pollution_score  greenhouse_gas_score  
count  2467.000000          2467.000000           2467.000000  
mean     21.114714             5.565464              5.072558  
std       5.745287             1.766611              1.765214  
min      10.000000             1.000000              0.000000  
25%      17.000000             5.000000              4.000000  
50%      21.000000             6.000000              5.000000  
75%      24.000000             7.000000              6.000000  
max     108.000000            10.000000             10.000000

Ignore the year there, it's supposed to be categorical but there are only two years there anyways. GHG_score has a value of zero(0), and I guess they won't be smartway certified. There seems to be large numbers in the city_mpg(124), hwy_mpg(99) and cmb_mpg(108), we'll have a look at the rows with high cmb_mpg,and also observations with zero(0) GHG_score. 

In [80]:
merged_df[merged_df['city_mpg'] > 50]

model       make  year    displ  cyl  \
1878                     CHEVROLET Volt  CHEVROLET  2018  1.50000    4   
1879                     CHEVROLET Volt  CHEVROLET  2018  1.50000    4   
1954  FORD Fusion Energi Plug-in Hybrid       FORD  2018  2.00000    4   
2150                  KIA Soul Electric        KIA  2018  3.04976    5   
2410             TOYOTA Camry Hybrid LE     TOYOTA  2018  2.50000    4   

       trans trans_type drive                  fuel      veh_class  city_mpg  \
1878     CVT        CVT   2WD  Gasoline/Electricity      small car        78   
1879     CVT        CVT   2WD  Gasoline/Electricity      small car        78   
1954     CVT        CVT   2WD  Gasoline/Electricity    midsize car        72   
2150  Auto-1       Auto   2WD           Electricity  station wagon       124   
2410   SCV-6        SCV   2WD              Gasoline    midsize car        51   

      hwy_mpg  cmb_mpg  air_pollution_score  greenhouse_gas_score smartway  
1878       70       74                    3                    10      Yes  
1879       70       74                    7                    10    Elite  
1954       66       69                    7                    10    Elite  
2150       93      108                   10                    10    Elite  
2410       53       52                    7                    10    Elite

I just saw that some values in disp has more than 2 decimal places. I have to round the column to 2dp.
The high mpg cars seem to be those with electricity fuel types, so they seem okay to keep. Then looking at obervations with zero(0) GHG_score

In [81]:
merged_df['displ'] = merged_df['displ'].round(2)

In [82]:
#for GHG_scores with values of zero
merged_df[merged_df['greenhouse_gas_score']==0]


model           make  year  displ  cyl  \
14                   ASTON MARTIN DB9          ASTON  2008    5.9   12   
16           ASTON MARTIN DB9 Volante          ASTON  2008    5.9   12   
63                     BENTLEY Arnage        BENTLEY  2008    6.7    8   
64                     BENTLEY Arnage        BENTLEY  2008    6.7    8   
65                      BENTLEY Azure        BENTLEY  2008    6.7    8   
66    BENTLEY Continental Flying Spur        BENTLEY  2008    6.0   12   
68            BENTLEY Continental GTC        BENTLEY  2008    6.0   12   
167                    BUGATTI Veyron        BUGATTI  2008    8.0   16   
497           FERRARI 599 GTB Fiorano        FERRARI  2008    5.9   12   
498           FERRARI 599 GTB Fiorano        FERRARI  2008    5.9   12   
499            FERRARI 612 Scaglietti        FERRARI  2008    5.7   12   
500            FERRARI 612 Scaglietti        FERRARI  2008    5.7   12   
828               JEEP Grand Cherokee           JEEP  2008    6.1    8   
889       LAMBORGHINI Gallardo Spyder    LAMBORGHINI  2008    5.0   10   
890       LAMBORGHINI Gallardo Spyder    LAMBORGHINI  2008    5.0   10   
891            LAMBORGHINI Murcielago    LAMBORGHINI  2008    6.5   12   
892            LAMBORGHINI Murcielago    LAMBORGHINI  2008    6.5   12   
893   LAMBORGHINI Murcielago Reventon    LAMBORGHINI  2008    6.5   12   
894   LAMBORGHINI Murcielago Roadster    LAMBORGHINI  2008    6.5   12   
895   LAMBORGHINI Murcielago Roadster    LAMBORGHINI  2008    6.5   12   
1046            MERCEDES-BENZ G55 AMG  MERCEDES-BENZ  2008    5.4    8   
1057         MERCEDES-BENZ Maybach 57  MERCEDES-BENZ  2008    5.5   12   
1058        MERCEDES-BENZ Maybach 57S  MERCEDES-BENZ  2008    6.0   12   
1059         MERCEDES-BENZ Maybach 62  MERCEDES-BENZ  2008    5.5   12   
1060        MERCEDES-BENZ Maybach 62S  MERCEDES-BENZ  2008    6.0   12   

        trans trans_type drive      fuel    veh_class  city_mpg  hwy_mpg  \
14      Man-6        Man   2WD  Gasoline    small car        10       16   
16      Man-6        Man   2WD  Gasoline    small car        10       16   
63    Auto-S6       Auto   2WD  Gasoline    large car         9       15   
64    Auto-S6       Auto   2WD  Gasoline  midsize car        10       15   
65    Auto-S6       Auto   2WD  Gasoline    small car         9       15   
66    Auto-S6       Auto   4WD  Gasoline  midsize car        10       17   
68    Auto-S6       Auto   4WD  Gasoline    small car        10       17   
167   Auto-S6       Auto   4WD  Gasoline    small car         8       14   
497    Auto-6       Auto   2WD  Gasoline    small car        11       15   
498     Man-6        Man   2WD  Gasoline    small car        11       15   
499    Auto-6       Auto   2WD  Gasoline  midsize car         9       16   
500     Man-6        Man   2WD  Gasoline  midsize car        10       15   
828   Auto-L5       Auto   4WD  Gasoline          SUV        11       14   
889     Man-6        Man   4WD  Gasoline    small car        10       15   
890   Auto-S6       Auto   4WD  Gasoline    small car        10       16   
891     Man-6        Man   4WD  Gasoline    small car         8       13   
892   Auto-S6       Auto   4WD  Gasoline    small car         9       14   
893   Auto-S6       Auto   4WD  Gasoline    small car         9       14   
894     Man-6        Man   4WD  Gasoline    small car         8       13   
895   Auto-S6       Auto   4WD  Gasoline    small car         9       14   
1046  Auto-L5       Auto   4WD  Gasoline          SUV        11       13   
1057  Auto-L5       Auto   2WD  Gasoline    large car        10       16   
1058  Auto-L5       Auto   2WD  Gasoline    large car        10       16   
1059  Auto-L5       Auto   2WD  Gasoline    large car        10       16   
1060  Auto-L5       Auto   2WD  Gasoline    large car        10       16   

      cmb_mpg  air_pollution_score  greenhouse_gas_score smartway  
14         12                    6                     0      

In [83]:
#checking unigue fuel types
merged_df['fuel'].unique()

array(['Gasoline', 'ethanol/gas', 'ethanol', 'CNG', 'diesel',
       'Gasoline/Electricity', 'Diesel', 'Ethanol/Gas', 'Electricity'],
      dtype=object)

In [84]:
#converting all to title case
merged_df['fuel'] = merged_df['fuel'].str.title()
merged_df['fuel'].unique()

array(['Gasoline', 'Ethanol/Gas', 'Ethanol', 'Cng', 'Diesel',
       'Gasoline/Electricity', 'Electricity'], dtype=object)

In [85]:
merged_df['make'].nunique()

52

In [86]:
merged_df['veh_class'].unique()

array(['SUV', 'midsize car', 'small car', 'station wagon', 'large car',
       'minivan', 'pickup', 'van', 'small SUV', 'standard SUV',
       'special purpose'], dtype=object)

In [87]:
merged_df['smartway'].unique()

array(['no', 'yes', 'No', 'Yes', 'Elite'], dtype=object)

In [88]:
merged_df['smartway']=merged_df['smartway'].str.title()
merged_df['smartway'].unique()

array(['No', 'Yes', 'Elite'], dtype=object)

In [89]:
merged_df[merged_df['fuel'] == 'Electricity']

model make  year  displ  cyl   trans trans_type drive  \
2150  KIA Soul Electric  KIA  2018   3.05    5  Auto-1       Auto   2WD   

             fuel      veh_class  city_mpg  hwy_mpg  cmb_mpg  \
2150  Electricity  station wagon       124       93      108   

      air_pollution_score  greenhouse_gas_score smartway  
2150                   10                    10    Elite

In [90]:
merged_df.to_csv('merged_fuel_economy_data.csv', index = False)

In [91]:
df = pd.read_csv('merged_fuel_economy_data.csv')
df.head()

model   make  year  displ  cyl    trans trans_type drive      fuel  \
0  ACURA MDX  ACURA  2008    3.7    6  Auto-S5       Auto   4WD  Gasoline   
1  ACURA MDX  ACURA  2008    3.7    6  Auto-S5       Auto   4WD  Gasoline   
2  ACURA RDX  ACURA  2008    2.3    4  Auto-S5       Auto   4WD  Gasoline   
3  ACURA RDX  ACURA  2008    2.3    4  Auto-S5       Auto   4WD  Gasoline   
4   ACURA RL  ACURA  2008    3.5    6  Auto-S5       Auto   4WD  Gasoline   

     veh_class  city_mpg  hwy_mpg  cmb_mpg  air_pollution_score  \
0          SUV        15       20       17                    7   
1          SUV        15       20       17                    6   
2          SUV        17       22       19                    7   
3          SUV        17       22       19                    6   
4  midsize car        16       24       19                    7   

   greenhouse_gas_score smartway  
0                     4       No  
1                     4       No  
2                     5       No  
3                     5       No  
4                     5       No